### PR TITLE
[MAX] Propose to Add LTX-2's Audio Autoencoder

### PR DESCRIPTION
**Pull request by tolgacangoz**
## Summary

This pull request introduces support for a new audio autoencoder architecture and adds a dropout module to the neural network package. The main changes include the implementation and registration of the `AutoencoderKLLTX2AudioModel`.

## Testing

<details>
<summary>Testing Code</summary>

```py
import json
import os
import subprocess
import sys
from pathlib import Path
from unittest.mock import MagicMock
import torch
import numpy as np
from typing import Any, Dict

# ==============================================================================
# 0. Modal Environment Setup (Optional/Auto-detected)
# ==============================================================================
# These are the settings used on Modal. If running elsewhere, they can be ignored.
MODEL_CLONE_PATH = "/root/modular"
BRANCH = "add-ltx-2-audio-vae"
REMOTE = "https://github.com/tolgacangoz/modular.git"

os.environ["HF_HOME"] = "/mnt/z-image/huggingface"
os.environ["MODULAR_HOME"] = "/mnt/z-image/modular"
os.environ["MODULAR_MAX_DEBUG"] = "True"

# Clone or update the repo first, before touching sys.path or sys.modules.
if not Path(MODEL_CLONE_PATH).exists():
    subprocess.run(["git", "clone", "--branch", BRANCH, REMOTE, MODEL_CLONE_PATH], check=True)
else:
    subprocess.run(["git", "-C", MODEL_CLONE_PATH, "switch", BRANCH], check=True)
    subprocess.run(["git", "-C", MODEL_CLONE_PATH, "pull"], check=True)

# Insert the cloned source at the front of sys.path.
path_to_max = os.path.join(MODEL_CLONE_PATH, "max/python")
if path_to_max not in sys.path:
    sys.path.insert(0, path_to_max)

# Evict any pre-cached max.* entries (Modal pre-imports the installed package).
# This forces a fresh search from sys.path on the next import.
for _mod in list(sys.modules):
    if _mod == "max" or _mod.startswith("max."):
        del sys.modules[_mod]

# Mock the generated protobuf modules missing from the source tree.
sys.modules["max.serve.kvcache_agent.kvcache_agent_service_v1_pb2"] = MagicMock()
sys.modules["max.serve.kvcache_agent.kvcache_agent_service_v1_pb2_grpc"] = MagicMock()

# ==============================================================================
# 1. Imports (from MAX and Diffusers)
# ==============================================================================
from huggingface_hub import hf_hub_download
from max.pipelines.architectures.autoencoders import AutoencoderKLLTX2AudioModel
from max.graph.weights import load_weights
from max.dtype import DType
from max.driver import Accelerator, CPU, accelerator_count
from max.experimental import Tensor
from diffusers import AutoencoderKLLTX2Audio

LATENT_DOWNSAMPLE_FACTOR = 4

# ==============================================================================
# 2. Equivalence Test Logic
# ==============================================================================
def convert_diffusers_config_to_max(diff_config: Dict[str, Any]) -> Dict[str, Any]:
    """Converts a Diffusers VAE config to MAX format."""
    max_cfg = dict(diff_config)
    if "output_channels" in max_cfg:
        max_cfg["out_channels"] = max_cfg.pop("output_channels")
    if "dtype" not in max_cfg:
        max_cfg["dtype"] = "bfloat16"
    return max_cfg

def test_equivalence():
    device = Accelerator() if accelerator_count() > 0 else CPU()
    print(f"Using device: {device}")

    repo_id, subfolder = "Lightricks/LTX-2", "audio_vae"
    hf_home = os.environ.get("HF_HOME")

    # 1. Download config and weights natively (cached after first run)
    print(f"Downloading config and weights from {repo_id}/{subfolder}...")
    config_path = hf_hub_download(
        repo_id, f"{subfolder}/config.json", cache_dir=hf_home
    )
    weights_path = hf_hub_download(
        repo_id,
        f"{subfolder}/diffusion_pytorch_model.safetensors",
        cache_dir=hf_home,
    )
    with open(config_path) as f:
        cfg = convert_diffusers_config_to_max(json.load(f))

    # 2. Initialize MAX Model with native SafetensorWeights
    print("Initializing MAX model with native weights...")
    max_model = AutoencoderKLLTX2AudioModel(
        config=cfg,
        encoding="bfloat16",
        devices=[device],
        weights=load_weights([Path(weights_path)]),
    )

    # 3. Load Diffusers model as reference oracle only
    print(f"Loading Diffusers reference model from {repo_id}/{subfolder}...")
    diff_model = AutoencoderKLLTX2Audio.from_pretrained(
        repo_id,
        subfolder=subfolder,
        torch_dtype=torch.bfloat16,
    ).to('cuda').eval()

    # 4. Compile the Fused Graph
    #
    # PostprocessAndDecode takes patchified latents of shape (B, T, D) where
    #   D = latent_channels * latent_mel_bins  (e.g. 8 * 16 = 128)
    # This mirrors the diffusers pipeline which runs:
    #   _denormalize_audio_latents(latents, ...)  ← on patchified (B, T, D)
    #   _unpack_audio_latents(latents, ...)       ← to (B, C, T, M)
    #   audio_vae.decode(latents)
    print("Compiling MAX fused decode graph...")
    latent_channels = cfg.get("latent_channels", 8)          # 8
    latent_mel_bins = cfg["mel_bins"] // LATENT_DOWNSAMPLE_FACTOR  # 64//4 = 16
    D = latent_channels * latent_mel_bins                    # 128

    fused_forward = max_model.build_fused_decode(device, num_channels=D, latent_mel_bins=latent_mel_bins)

    # 4. Prepare Realistic Inputs
    #
    # Input is patchified audio latents (B, T, D):
    #   T = latent audio frames (126 from the denoising loop)
    #   D = 128 (= latent_channels * latent_mel_bins)
    #
    # Reference shapes from the denoising loop (for context):
    #   audio_latent_model_input: torch.Size([2, 126, 128])
    #   audio_vae's latents (unpacked): torch.Size([1, 8, 126, 16])
    print("Generating test input (patchified latents)...")
    batch_size = 1
    T = 126  # latent audio frames
    z_patchified = torch.randn(batch_size, T, D, dtype=torch.bfloat16).to('cuda')

    # 5. Run Equivalence Test
    print("Running forward pass...")

    # MAX: PostprocessAndDecode handles denorm + unpack + decode internally
    max_input = Tensor.from_dlpack(z_patchified).to(device)
    max_output = fused_forward(max_input)
    max_output_np = np.from_dlpack(max_output.cast(DType.float32).to(CPU()))

    # Diffusers equivalent (mirrors the pipeline's decode path exactly):
    #   1. _denormalize_audio_latents: broadcast (D,) stats over (B, T, D)
    #   2. _unpack_audio_latents: (B, T, D) → (B, C, T, M)
    #   3. audio_vae.decode
    std = diff_model.latents_std   # shape (D,) = (128,)
    mean = diff_model.latents_mean # shape (D,) = (128,)
    z_denormed = z_patchified * std + mean                                     # (B, T, D)
    z_unpacked = z_denormed.unflatten(-1, (latent_channels, latent_mel_bins))  # (B, T, C, M)
    z_unpacked = z_unpacked.permute(0, 2, 1, 3)                               # (B, C, T, M)

    with torch.no_grad():
        diff_output = diff_model.decode(z_unpacked, return_dict=False)[0]
    diff_output_np = diff_output.detach().float().cpu().numpy()

    # Final Comparison
    print(f"Input shape  (patchified): {list(z_patchified.shape)}")
    print(f"Diffusers output shape:    {list(diff_output.shape)}")
    print(f"MAX output shape:          {list(max_output_np.shape)}")

    abs_diff = np.abs(max_output_np - diff_output_np)
    n_total = abs_diff.size
    print(f"Max absolute difference:   {abs_diff.max():.6f}")
    print(f"Mean absolute difference:  {abs_diff.mean():.6f}")
    print(f"Median absolute difference:{np.median(abs_diff):.6f}")
    # bfloat16 ULP at the output scale (~4-8): 0.03125 per ULP.
    # Acceptable threshold for a full bfloat16 decode: 4 ULPs = 0.125.
    BF16_ATOL = 0.125
    BF16_RTOL = 0.02
    n_fail = int(np.sum(
        abs_diff > BF16_ATOL + BF16_RTOL * np.abs(diff_output_np)
    ))
    print(f"Elements outside {BF16_ATOL} + {BF16_RTOL}*|ref|: {n_fail}/{n_total} ({100*n_fail/n_total:.1f}%)")

    try:
        np.testing.assert_allclose(max_output_np, diff_output_np, rtol=BF16_RTOL, atol=BF16_ATOL)
        print("SUCCESS: MAX and Diffusers VAE outputs are equivalent (within bfloat16 tolerance)!")
    except AssertionError as e:
        print(f"FAILURE: Outputs differ beyond bfloat16 tolerance. Error: {e}")

test_equivalence()

"""
Using device: Device(type=gpu,id=0)
Downloading config and weights from Lightricks/LTX-2/audio_vae...
Initializing MAX model with native weights...
Loading Diffusers reference model from Lightricks/LTX-2/audio_vae...
Compiling MAX fused decode graph...
Generating test input (patchified latents)...
Running forward pass...
Input shape  (patchified): [1, 126, 128]
Diffusers output shape:    [1, 2, 501, 64]
MAX output shape:          [1, 2, 501, 64]
Max absolute difference:   0.062500
Mean absolute difference:  0.007654
Median absolute difference:0.000000
Elements outside 0.125 + 0.02*|ref|: 0/64128 (0.0%)
SUCCESS: MAX and Diffusers VAE outputs are equivalent (within bfloat16 tolerance)!
"""
```

</details>

## Checklist

- [X] PR is small and focused — consider splitting larger changes into a
      sequence of smaller PRs
- [X] I ran `./bazelw run format` to format my changes
- [X] I added or updated tests to cover my changes
- [X] If AI tools assisted with this contribution, I have included an
      `Assisted-by:` trailer in my commit message or this PR description
      (see [AI Tool Use Policy](../AI_TOOL_POLICY.md))

Assisted-by: AI

@KCaverly @katelyncaldwell

### CODE REVIEW

File: max/python/max/experimental/nn/dropout.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+# ===----------------------------------------------------------------------=== #
+"""A Module for dropout layers."""
+
+from __future__ import annotations
+
+from max.experimental.nn import Module
+from max.experimental.tensor import Tensor
+
+
+class Dropout(Module[[Tensor], Tensor]):
```
Comment (tolgacangoz) [inline max/python/max/experimental/nn/dropout.py:22]: I can remove this if it seems inappropriate since `modular` doesn't support training officially at the moment. I added it just to emphasize easy-to-migrability from `diffusers` and/or `torch` without doing parameter name arrangements.

**Comment by tolgacangoz** [comment]
I cannot take another risk due to another potential **internal** arrangement, thus I must leave LTX-2 PRs as mostly completed.

I recommend continuing LTX-2 from these mostly completed PRs and request adding me as a coauthor.

Thanks for the collaboration on my merged, discussed PRs so far.

@KCaverly @katelyncaldwell
@jglee-sqbits @pei0033 @byungchul-sqzb @kkimmk @tae-su-kim
